### PR TITLE
Changing generic weight of `tyGenericParam`

### PR DESCRIPTION
**Pull request by Graveflo**
This is in reference to a [feature request](https://github.com/nim-lang/Nim/issues/22142) that I posted.

I'm making this PR to demonstrate the suggested change and expect that this should be scrutinized

### CODE REVIEW

**Comment by Graveflo** [comment]
On the CI failure, there is a test in the Nim codebase under `tests/macros/tgetimpl` that passes `len` as a typed parameter to a macro and then uses the expression  `y.owner` where the parameter is `y`, as if `len` will not have other candidates. I'm not sure why this compiles with the current version in nim. It is given the `nnkSym` node kind in current nim and `nnkClosedSymChoice` in the PR'd version of nim. I'm still looking into the datamancer failure

Also, I know the test file is a little bit messed up. If preferable I can get rid of some of those "anti-regression" tests and I will fix the `while true` block inside of the `block` that I made.

Also, if this ends up working I think I should add a blurb to the manual

**Comment by Araq** [comment]
> Also, if this ends up working I think I should add a blurb to the manual

You really should start with the manual/spec.

**Comment by Graveflo** [comment]
Is there a formal spec? I looked but all I could find was the manual.
I'll add some text to the manual that talks about some of the pieces involved in this, as I think that clarification would be helpful in multiple ways.
To be clear, I don't think that I am trying to do make nim behave off-spec as it is described in the manual. It is a breaking change, but not contradicting the manual as far as I know. The test cases (barring the "anti regression" mess) are good isolated examples of what nim currently does vs what I assume it should do.
Also, I want to make sure I have my terminology correct here, because this should be important when explaining things in the manual. Is this accurate for the "anatomy of a call":
```nim
p[T: t](arg1: f): y
```
`p`: callee symbol

`[...]`: constraints

`T: t`: constraint

`(arg1: f): y`: formal signature

`arg1: f`: formal parameters

`f`: formal parameter type

`y`: formal return type


After matching / resolution the "formal" parameters that are not concrete crystalize into call-specific bindings.

**Comment by Araq** [comment]
> Is there a formal spec? I looked but all I could find was the manual.

The manual is the spec. Could be more formal, we're getting there by telling contributors to write spec'y documentation.

**Comment by Araq** [comment]
It is:

`p`: callee symbol

`[...]`: generic parameters

`T: t`: generic constraint

`[T: t](arg1: f): y`: (formal) signature

`arg1: f`: (formal) parameter

`f`: (formal) parameter type

`y`: (formal) return type

**Comment by Graveflo** [comment]
Alright, thank you and `T` is described as a "type variable" in the manual. This issue was a bit more complex then I realized.
I am going to:
- Give the manual another look and see if there is more I can clarify there
- Look at the datamancer failure in the CI but it is a web of macros that is hard to parse
- Change the `tgetimpl` test to use something other then `len` because it has more then one candidate, so I think `owner` doesn't make sense there
- Maybe make another issue with a condensed version of the bug in `tgetimpl` and make test cases for it, although I have no idea why it is currently accepted by the compiler
- Look into other type classes in `typeRel` because I only modified how `object` behaves there and other type classes might need similar treatment

If I can iron out that list I might be done with this. For a first pass at least.

If anyone thinks that test in `tgetimpl` should actually work or has info about the datamancer failure feel free to let me know

**Comment by Graveflo** [comment]
Currently, this is a minimal example that demonstrates a breakage in this PR:
```nim
import sugar

proc p[T]()=
  var t = collect(newSeq): 1
p[int]()
```
The output is `Error: Expected one of {nnkIdent, nnkSym, nnkCall}, got nnkOpenSymChoice` similar to the `tgetimpl` breakage.
```
compiler/semstmts.nim(662) semVarOrLet
compiler/semexprs.nim(119) semExprWithType
compiler/semexprs.nim(78) semExprCheck
compiler/semexprs.nim(3146) semExpr
compiler/semexprs.nim(1147) semDirectOp
compiler/semexprs.nim(1025) afterCallActions
compiler/sem.nim(502) semMacroExpr
compiler/vm.nim(2484) evalMacroCall
compiler/vm.nim(1899) rawExecute
compiler/vm.nim(75) stackTraceImpl
compiler/msgs.nim(579) liMessage
```
I think I may have inadvertently broken procs or some binding. I'll be looking into it

**Comment by Graveflo** [comment]
There are two failing CI cases left and I think they are out of scope:

**Datamancer**: This error seems to be caused the heuristic type inference in formulas. The logic inspects the node kind of symbols to check for values like `nnkClosedSymChoice` and `nnkSym` which are affected by overload resolution. This issue seems solvable by a less fragile approach. After looking into it I don't think this error implicates an issue with the PR on it's own.

**Neo**: The error message is clear. Two procs with `(int, int, ...)` signature and a call supplied with two `int` literals. I'm not sure why this previously compiled.

** Note: I disabled them due to above

**Comment by Graveflo** [comment]
See above for rational on broken packages.
CI failed this last go due to Mac server shitting out.

The changes in `sumGeneric` have the purpose of not counting "containers" and only counting substantial items. Generic params are now factored in. Also, `tyTypeDesc` has a `tyNone` generic parameter sometimes and that should not be counted.

In essence, there was a lot of logic that relied on `sumGeneric` incorrectly because `typeRel` would return `isNone` overzealously. Now `typeRel` is more accurate and doesn't need to rely on `sumGeneric` being overburdened.

`isObjectSubtype` needed slightly more robustness since it is used in more situations.

In `typeRel`:
* `tyArray`: Since `isGeneric` is now more common a check for concreteness is needed for static evaluation
* `tyObject`: There is a check for `tyAlias` to keep concepts working. Otherwise, the object type of the subject is checked for membership instead of just checking the kind
* `tyPtr, tyRef`: Alignmnet of these nodes can be off if buried in generic params or bodies. `getObjectTypeOrNil` is meant to mitigate this. It is also used in `tyBuiltInTypeClass`
* `tyGenericInvocation`: `tyGenericParam` is skipped
* `tyBuiltInTypeClass`: Same extraction routine of type class as in `tyPtr`, `tyRef` except we have to skip `tyBuiltInTypeClass` to line up with `f[0]`
* `tyGenericParam`: `doBind` is used to differentiate between signature comparisons and parameter matching. There is a special case for non-concrete values in binding mode

test cases have justifications in comments. This ended up having quite a bit of scope creep, but in the end, I think `typeRel` more accurately does what it is supposed to do now.

**Comment by metagn** [comment]
Made a PR to neo to fix CI https://github.com/andreaferretti/neo/pull/54, already merged into the master branch of the neo fork used in CI. You can re-enable neo and mention https://github.com/andreaferretti/neo/pull/54 additionally to the other PR link in the comment below it.

> After looking into it I don't think this error implicates an issue with the PR on it's own.

Can you explain what datamancer has to do to work? Unless you can show datamancer is wrong, the PR is at fault for its CI failure.

**Comment by Graveflo** [comment]
> Can you explain what datamancer has to do to work? Unless you can show datamancer is wrong, the PR is at fault for its CI failure.

Here is some code that will make the error happen in the PR version of nim:
```nim
import datamancer
discard f{`a` + `b`}
```
This can be traced like so `compileFormulaImpl` -> `determineTypes` -> `determineTypesImpl` (`of nnkInfix` branch) -> `findType`

There are case statements that observe node kinds like `nnkClosedSymChoice`, `nnkOpenSymChoice` and `nnkSym`. The problem that I have seen with this approach is that overload resolution affects how nim classifies the nodes. Datamancer is trying to guess what type everything should be from the operator (`+` in this case) but using `nnkClosedSymChoice` vs `nnkSym` for a symbol like `+` is not really directly testing for what is needed. It seems to me that this approach is relying on property that just so happens to work out in this situation for a particular version of nim. I assume that the symbol can simply be tested explicitly.  This code works because `+` is not affected:

```nim
import datamancer
discard f{`a` - `b`}
```

**Comment by Graveflo** [comment]
PR #23033 appears to fix datamancer on this PR. As far as I know that is the last caveat here, so this should be ready if/when that gets merged

File: doc/manual.md [hunk 1]
```diff
@@ -2669,10 +2670,12 @@ algorithm returns true:
   return "ambiguous"
   ```
 
-When counting is ambiguous, disambiguation begins. Parameters are iterated
-by position and these parameter pairs are compared for their type relation. The general goal
-of this comparison is to determine which parameter is more specific. The types considered are
-not of the inputs from the callsite, but of the competing candidates' parameters.
+When counting is ambiguous, disambiguation begins. Disamgiuation also has two stages, first a
```
Comment (Araq) [inline doc/manual.md]:
```suggestion
When counting is ambiguous, disambiguation begins. Disambiguation also has two stages, first a
```

File: doc/manual.md [hunk 2]
```diff
@@ -2734,6 +2736,23 @@ matches) is preferred:
...
+For example:
+```nim
+type A
+proc p[T: A](param: T)
+proc p[T: object](param: T)
+```
+
+These signatures are not ambiguous for an instance of `A` even though the formal parameters match ("T" == "T").
+Instead `T` is treated as a variable in that (`T` ?= `T`) depending on the bound type of `T` at the time of
+overload resoliution.
```
Comment (Araq) [inline doc/manual.md]:
```suggestion
overload resolution.
```

**Comment by Araq** [comment]
I'm merging this but it needs a follow PR as the documentation is incomprehensible. Also, the whole approach is kinda nuts (not your fault, of course). Instead we need a formal type relation `isMoreSpecific(a, b)` that describes the effects, not how it's implemented.

**Comment by Graveflo** [comment]
> we need a formal type relation `isMoreSpecific(a, b)` that describes the effects, not how it's implemented.

Just for posterity and to throw an idea out there: The hypothetical function `isMoreSpecific` maps to `checkGeneric` in the compiler source. Specificity is a hierarchical rank. `typeRel` defines a hierarchy implicitly and `checkGeneric` enforces it explicitly in overload resolution, so these two functions should be the same.

Documenting "specificity" would be 99% documenting `typeRel` which is an interesting prospect considering as how complex it is, and it's current documentation status. There's more that can be said about this but I wouldn't go into further detail with this prompting.

Also, for the record, I don't like the way specificity sounds in these contexts. I'm the one that started saying it, so not blaming anyone but me for that. Still, it does seem to fit the idea well enough.

#23171
